rfcs/0042-config-option.md

```diff
@@ -0,0 +1,350 @@
+    check-interval = mkIf isServer 3600;
+  };
+}
+```
+
+See [here](#an-example) for an example for how a NixOS module will look like.
+
+Jump to the [detailed design][part1-design] of part 1
+
+## Part 2: Balancing module option count
+[part2]: #part-2-balancing-module-option-count
+
+Since with this approach there will be no more hardcoded defaults and composability is not a problem anymore, there is not a big need to have NixOS options for every setting anymore. Traditionally this has lead to huge modules with dozens of options, each of them only for a single field in the configuration. Such modules are problematic because they're hard to write, review and maintain, are generally of lower quality, fill the option listing with noise and more. Additional options aren't without advantages however: They are presented in the NixOS manual and can have better type checking than the equivalent with `settings`.
+
+The second part of this RFC aims to encourage module authors to strike a balance for the number of additional options such as to not make the module too big, but still provide the most commonly used settings as separate options. Quality is encouraged over quantity: Authors should spend more time on writing documentation, NixOS tests or useful high-level abstractions. This is in contrast to the fiddly labor of copying dozens of options from upstream to NixOS. With a `settings` option, it's also very easy to add additional options over time if the need arises. In contrast, removing options has always been nigh impossible.
+
+Jump to the [detailed design][part2-design] of part 2
+
+# Motivation
+[motivation]: #motivation
+
+## [Part 1][part1]
+
+Stringly-typed options such as `extraConfig` have multiple disadvantages in comparison to a structural `settings` option.
+- Impossible to even implement correctly with configuration formats like JSON (because concatenation doesn't make sense)
+- Bad modularity
+  - No proper merging: Multiple assignments get merged together with string concatenation which can't merge assignments of the same setting
+  - No priorities: `mkDefault` and co. won't work on settings
+- Values within it can't be inspected, since it's just an opaque string
+- Syntax of assigned values can easily be wrong, especially with escaping sequences
+- Can break services if users assign a value to `extraConfig` which later gets turned into a specialized option, [here](https://github.com/NixOS/nixpkgs/commit/23d1c7f4749#diff-d66632f8013e5976f782de43a0043604R750) is an example of this.
+
+## [Part 2][part2]
+
+NixOS modules with dozens of options aren't optimal for these reasons:
+- Writing takes a lot of time and is a repetitive task of copying the settings from upstream
+- Reviewing is tedious because of the big size of it, which in turn demotivates reviewers to even do it
+- The option listing will be filled with a lot of options almost nobody ever needs, which in turn makes it hard for people to find the options they *do* need.
```

Comment:
A good exemplification of a technically failed project (though still hugely popular) due to vast amounts of input variables is helm charts.